File: MvvmSample/app/src/main/java/com/abben/mvvmsample/ui/movies/MoviesFragment.java
package com.abben.mvvmsample.ui.movies;

import android.content.Intent;
import android.os.Bundle;
import android.util.DisplayMetrics;
import android.view.View;

import com.abben.mvvmsample.R;
import com.abben.mvvmsample.base.BaseDatabindFragment;
import com.abben.mvvmsample.databinding.FragmentMoviesBinding;
import com.abben.mvvmsample.ui.moviedetail.MovieDetailsActivity;
import com.abben.mvvmsample.ui.movies.adapter.MoviesRecycleViewAdapter;
import com.abben.mvvmsample.vo.TypeMovies;

import androidx.core.app.ActivityOptionsCompat;
import androidx.lifecycle.ViewModelProviders;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import static com.abben.mvvmsample.MainActivity.INTENT_MOVIE_FALG;

/**
 * Created by abben on 2017/11/20.
 */
public class MoviesFragment extends BaseDatabindFragment<FragmentMoviesBinding> implements SwipeRefreshLayout.OnRefreshListener {
    private static String MOVE_TYPE_KEY = "moveType";

    private MoviesRecycleViewAdapter moviesRecycleViewAdapter;
    private MoviesViewModel moviesViewModel;

    public static MoviesFragment create(@TypeMovies.TypeMoviesAnnotation String moveType) {
        MoviesFragment moviesFragment = new MoviesFragment();
        Bundle bundle = new Bundle();
        bundle.putString(MOVE_TYPE_KEY, moveType);
        moviesFragment.setArguments(bundle);
        return moviesFragment;
    }

    @Override
    protected int getContentViewId() {
        return R.layout.fragment_movies;
    }

    @Override
    protected void initView() {
        viewBinding.fragmentSwipeRefresh.setOnRefreshListener(this);
        DisplayMetrics dm = getResources().getDisplayMetrics();
        moviesRecycleViewAdapter = new MoviesRecycleViewAdapter(getContext(), dm.widthPixels);
        moviesRecycleViewAdapter.setOnItemClikeListen((movie, view) -> {
            Intent intent = new Intent(getActivity(), MovieDetailsActivity.class);
            intent.putExtra(INTENT_MOVIE_FALG, movie);
            ActivityOptionsCompat optionsCompat = ActivityOptionsCompat.makeSceneTransitionAnimation(
                    getActivity(), view, getString(R.string.transitions_name));
            startActivity(intent, optionsCompat.toBundle());
        });
        viewBinding.fragmentRecyclerview.setAdapter(moviesRecycleViewAdapter);

        moviesViewModel = ViewModelProviders.of(this).get(MoviesViewModel.class);
        viewBinding.setRetryCallback(() ->
                moviesViewModel.retry()
        );
        moviesViewModel.setMoveType(getArguments().getString(MOVE_TYPE_KEY));
        subscribeUi(moviesViewModel);
    }

    @Override
    protected void initListener() {

    }

    @Override
    public void onClick(View view) {

    }

    /**
     * 绑定ViewModel与View,
     * 当数据改变时通知View
     *
     * @param moviesViewModel ViewModel
     */
    private void subscribeUi(MoviesViewModel moviesViewModel) {
        moviesViewModel.getmObservableMovies().observe(this, listResource -> {
            viewBinding.setRepoResource(listResource);
            if (listResource != null && listResource.data != null) {
                moviesRecycleViewAdapter.setMovies(listResource.data);
            }
            viewBinding.executePendingBindings();
        });
        moviesViewModel.getmRefreshing().observe(this, aBoolean -> {
            if (aBoolean != null) {
                viewBinding.fragmentSwipeRefresh.setRefreshing(aBoolean);
            }
        });
    }

    @Override
    public void onRefresh() {
        moviesViewModel.retry();
    }

}

File: MvvmSampleJetpack/app/src/main/java/com/abben/mvvmsamplejetpack/ui/movies/MoviesFragment.kt
package com.abben.mvvmsamplejetpack.ui.movies

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.core.app.ActivityOptionsCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.abben.mvvmsamplejetpack.MainActivity.Companion.INTENT_MOVIE_FALG
import com.abben.mvvmsamplejetpack.R
import com.abben.mvvmsamplejetpack.base.BaseDatabindFragment
import com.abben.mvvmsamplejetpack.bean.Movie
import com.abben.mvvmsamplejetpack.common.OnItemClickListener
import com.abben.mvvmsamplejetpack.common.RetryCallback
import com.abben.mvvmsamplejetpack.databinding.FragmentMoviesBinding
import com.abben.mvvmsamplejetpack.ui.moviedetail.MovieDetailsActivity
import com.abben.mvvmsamplejetpack.ui.movies.adapter.MoviesRecycleViewAdapter
import com.abben.mvvmsamplejetpack.vo.TypeMovies

/**
 *  @author abben
 *  on 2019/1/21
 */
class MoviesFragment : BaseDatabindFragment<FragmentMoviesBinding>()
    , SwipeRefreshLayout.OnRefreshListener {

    private var moviesRecycleViewAdapter: MoviesRecycleViewAdapter? = null
    private var moviesViewModel: MoviesViewModel? = null

    companion object {
        private const val MOVE_TYPE_KEY = "moveType"

        fun create(@TypeMovies moveType: String): MoviesFragment {
            val moviesFragment = MoviesFragment()
            val bundle = Bundle()
            bundle.putString(MOVE_TYPE_KEY, moveType)
            moviesFragment.arguments = bundle
            return moviesFragment
        }
    }

    override fun getContentViewId(): Int {
        return R.layout.fragment_movies
    }

    override fun initView() {
        viewBinding.fragmentSwipeRefresh.setOnRefreshListener(this)
        val dm = resources.displayMetrics
        moviesRecycleViewAdapter = MoviesRecycleViewAdapter(dm.widthPixels / 2 - 20)
        moviesRecycleViewAdapter!!.mOnItemClickListener = object : OnItemClickListener<Movie> {
            override fun onItemClick(t: Movie, view: View) {
                val intent = Intent(activity, MovieDetailsActivity::class.java)
                intent.putExtra(INTENT_MOVIE_FALG, t)
                val optionsCompat = ActivityOptionsCompat.makeSceneTransitionAnimation(
                    activity!!, view, getString(R.string.transitions_name)
                )
                startActivity(intent, optionsCompat.toBundle())
            }
        }


        viewBinding.fragmentRecyclerview.adapter = moviesRecycleViewAdapter

        moviesViewModel = ViewModelProviders.of(this).get(MoviesViewModel::class.java)
        viewBinding.retryCallback = object : RetryCallback {
            override fun retry() {
                moviesViewModel?.retry()
            }
        }
        moviesViewModel?.moveType?.value = arguments?.getString(MOVE_TYPE_KEY)
        subscribeUi(moviesViewModel)
    }

    override fun initListener() {
    }

    override fun onClick(p0: View?) {
    }

    override fun onRefresh() {
        moviesViewModel?.retry()
    }

    /**
     * 绑定ViewModel与View,
     * 当数据改变时通知View
     *
     * @param moviesViewModel ViewModel
     */
    private fun subscribeUi(moviesViewModel: MoviesViewModel?) {
        moviesViewModel?.mObservableMovies?.observe(this, Observer {
            viewBinding.repoResource = it
            it.data?.let { movies ->
                moviesRecycleViewAdapter?.setMovies(movies)
            }
            viewBinding.executePendingBindings()
        })
        moviesViewModel?.getmRefreshing()?.observe(this, Observer {
            viewBinding.fragmentSwipeRefresh.isRefreshing = it
        })
    }
}